nixos/modules/services/network-filesystems/tahoe.nix

```diff
@@ -0,0 +1,175 @@
+            default = 3;
+            type = types.int;
+            description = ''
+              The number of shares required to reconstitute a file.
+            '';
+          };
+          client.shares.happy = mkOption {
+            default = 7;
+            type = types.int;
+            description = ''
+              The number of distinct storage nodes required to store
+              a file.
+            '';
+          };
+          client.shares.total = mkOption {
+            default = 10;
+            type = types.int;
+            description = ''
+              The number of shares required to store a file.
+            '';
+          };
+          storage.enable = mkEnableOption "storage service";
+          storage.reservedSpace = mkOption {
+            default = "1G";
+            type = types.str;
+            description = ''
+              The amount of filesystem space to not use for storage.
+            '';
+          };
+          helper.enable = mkEnableOption "helper service";
+          package = mkOption {
+            default = pkgs.tahoelafs;
+            defaultText = "pkgs.tahoelafs";
+            type = types.package;
+            example = literalExample "pkgs.tahoelafs";
+            description = ''
+              The package to use for the Tahoe LAFS daemon.
+            '';
+          };
+        };
+      };
+    };
+    config = mkIf (cfg.nodes != {}) {
+      environment = {
+        etc = flip mapAttrs' cfg.nodes (node: settings:
+          nameValuePair "tahoe-lafs/${node}.cfg" {
+            mode = "0444";
+            text = ''
+              # This configuration is generated by Nix. Edit at your own
+              # peril; here be dragons.
+
+              [node]
+              nickname = ${settings.nickname}
+              tub.port = ${toString settings.tub.port}
+
+              [client]
+              ${optionalString (settings.client.introducer != null)
+                "introducer.furl = ${settings.client.introducer}"}
+              ${optionalString (settings.client.helper != null)
+                "helper.furl = ${settings.client.helper}"}
+
+              shares.needed = ${toString settings.client.shares.needed}
+              shares.happy = ${toString settings.client.shares.happy}
+              shares.total = ${toString settings.client.shares.total}
+
+              [storage]
+              enabled = ${if settings.storage.enable then "true" else "false"}
+              reserved_space = ${settings.storage.reservedSpace}
+
+              [helper]
+              enabled = ${if settings.helper.enable then "true" else "false"}
+            '';
+          });
+        # Actually require Tahoe, so that we will have it installed.
+        systemPackages = flip mapAttrsToList cfg.nodes (node: settings:
+          settings.package
+        );
+      };
+      # Open up the firewall.
+      networking.firewall.allowedTCPPorts = flip mapAttrsToList cfg.nodes
```

Comment:
I don't think our general practice is to auto-modify `networking.firewall` when enabling services, instead getting people to manually open up ports separately, as a way to acknowledge that they do want the port externally accessible. I sort of prefer that model, and think it's pretty common in existing NixOS services. Seems like a bigger conversation to have here, but until we've sorted out the bigger question, it's probably best to just stay consistent with the convention and take this out.